[notnotmax] iP

notnotmax opened the pull request:
# Qwerty
> "I didn't say that." - Sun Tzu, *The Art of War*

Qwerty frees your soul of having to remember things you should but don't want to do. It's,
- text-based :book:
- easy to learn 🤔 
- ~~FAST~~ *NEGLIGIBLY* FASTER to use :snail:

All you need to do is,
1. Download it from [here](https://github.com/notnotmax/ip).
2. Double-click the jar file.
3. Add your tasks.
4. Let it manage your tasks for you.

And it is **FREE**! 😮 

## Features
- [X] Managing three types of tasks (todo, deadline, event)
- [X] Autosave every time a task is updated
- [ ] Doing your tasks for you (coming never)

If you are a Java programmer, you can look at this nice piece of code. Here's a `main` method that throws an `Exception`:
```java
public class Main {
    public static void main(String[] args) {
        throw new Exception();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

notnotmax committed:
Implement skeletal version

notnotmax committed:
Add ability to echo

notnotmax committed:
Add list command

notnotmax committed:
Add Task.java

notnotmax committed:
Rename methods to markAsDone and markAsNotDone

notnotmax committed:
Add ability to mark and unmark

notnotmax committed:
Add Todo.java, Deadline.java and Event.java

notnotmax committed:
Add todo, event and deadline commands

notnotmax committed:
Add testcases for text ui testing

notnotmax committed:
Add QwertyException.java

notnotmax committed:
Add error handling for missing arguments

notnotmax committed:
Add error handling for invalid index

notnotmax committed:
Add delete command

notnotmax committed:
Rewrite method names in camelCase

notnotmax committed:
Update testcases

notnotmax committed:
Separate parsing of main command and additional args

notnotmax committed:
Add method to get task details

notnotmax committed:
Change .gitignore to ignore savefile

notnotmax committed:
Add method to create and write to save

notnotmax committed:
Add method to load task list from file

notnotmax committed:
Change time arguments to use LocalDateTime

notnotmax committed:
Create TaskList.java

notnotmax committed:
Add Storage.java

notnotmax committed:
Add *Command.java files

notnotmax committed:
Add Parser.java

notnotmax committed:
Connect Qwerty.java to Parser and Command

notnotmax committed:
Add Ui.java

notnotmax committed:
Remove print statements from TaskList.java

notnotmax committed:
Integrate Ui into commands and parser

notnotmax committed:
Sort files into packages

notnotmax committed:
Merge gradle branch into master

notnotmax committed:
Add ParserTest.java and TodoCommandTest.java

notnotmax committed:
Add TaskListTest.java

notnotmax committed:
Add DeadlineCommandTest.java

notnotmax committed:
Update filepaths in build.gradle

notnotmax committed:
Add Javadoc comments

notnotmax committed:
Make code compliant with Java coding standard

notnotmax committed:
Add find command

notnotmax committed:
Merge branch 'branch-A-CodingStandard'

notnotmax committed:
Merge branch 'branch-Level-9'

beginner-web-developer commented on src/main/java/qwerty/Parser.java:
```diff
@@ -0,0 +1,85 @@
+    public static HashMap<String, String> parseArgs(String additionalArgs) {
+        HashMap<String, String> map = new HashMap<>();
+
+        //extract first argument separately because it has no identifier
```
Perhaps there can be a space before the comment starts to maintain consistency with the rest of the code?

notnotmax replied:
Good catch, added a space.

beginner-web-developer commented on src/main/java/qwerty/Storage.java:
```diff
@@ -0,0 +1,135 @@
+import qwerty.task.Task;
+import qwerty.task.Todo;
+
+public class Storage {
```
I think this is missing a header comment.

notnotmax replied:
Added one.

beginner-web-developer commented on src/main/java/qwerty/Storage.java:
```diff
@@ -0,0 +1,135 @@
+                    try {
+                        // Parse the string in each line and match it to a task type
+                        String[] args = scanner.nextLine().split("\\|");
+                        switch (args[0]) {
```
Perhaps you can consider extracting the logic into a private method, as right now it is a little confusing due to the deep indentation.

beginner-web-developer commented on src/test/java/qwerty/command/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,104 @@
+import qwerty.TaskList;
+import qwerty.Ui;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Maybe it would be good to expand the wildcard import to show only the methods that you need.

notnotmax replied:
Accidental move by IDE auto-import, will fix.

beginner-web-developer reviewed (commented):
Overall I like how your code is nicely OOPed, and it is mostly following the coding standards, except for a few nits.

notnotmax committed:
Add GUI using JavaFX

notnotmax committed:
Make code compliant with coding standard

notnotmax committed:
Make main code compliant with checkstyle

notnotmax committed:
Make test code compliant with checkstyle

jialegitgud commented on src/main/java/qwerty/QwertyException.java:
```diff
@@ -0,0 +1,11 @@
+ * This class encapsulates exceptions unique to the Qwerty chatbot.
+ * An example is missing or invalid arguments.
+ */
+public class QwertyException extends Exception {
```
Perhaps this exception class could be named less general to reflect its use cases?

jialegitgud commented on src/main/java/qwerty/Parser.java:
```diff
@@ -0,0 +1,85 @@
+        HashMap<String, String> map = new HashMap<>();
+
+        // extract first argument separately because it has no identifier
+        String[] argSplit = additionalArgs.split(" /", 2);
```
Perhaps could be named `argSplits` to reflect the plural nature of the variable?

jialegitgud reviewed (commented):
Overall, very clean use of OOP and followed all the naming conventions perfectly, there was nothing to nit-pick 👍

notnotmax committed:
Update build.gradle

notnotmax committed:
Add assertions to Storage.java

Storage.java contains important assumptions that should hold during
execution.

Adding assertions will help make the code safer.

Let's add assertions to verify that:
* The file to read from does exist after executing `file.exists()`
* The arguments of the private `createTask` method are non-null

notnotmax committed:
Improve code quality

Parser.java has array variables that do not have plural names,
while Qwerty.java has unnecessarily indented code.

Rewriting these portions to adhere to the code quality standard
will make the code easier to read.

Let's:
* Rename array variables to better reflect their array nature
* Use guard clauses to reduce deep nesting in Qwerty.java

The code is being rewritten this way in accordance with the
suggestions in the textbook.

notnotmax committed:
Highlight use of streams

notnotmax committed:
Merge pull request #1 from notnotmax/branch-A-Assertions

Add assertions

notnotmax committed:
Merge pull request #2 from notnotmax/branch-A-CodeQuality

Improve code quality

notnotmax committed:
Merge pull request #3 from notnotmax/branch-A-Streams

Highlight stream usage

notnotmax committed:
Add .yml file for CI

notnotmax committed:
Update testcases

notnotmax committed:
Add sort command

notnotmax committed:
Fix checkstyle

notnotmax committed:
Update documentation

notnotmax committed:
Add title to UI window

notnotmax committed:
Update GUI

notnotmax committed:
Make bye command close GUI

notnotmax committed:
Update README

notnotmax committed:
Update README
